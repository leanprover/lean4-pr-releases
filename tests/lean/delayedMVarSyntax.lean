/-!
# Syntax to create delayed assignment metavariables
-/

/-!
Creating and reusing some synthetic holes using delayed assignment syntax.
-/
example : True := by
  let f : Nat → Nat := fun x => ?a
  let g : Nat → Nat → Nat := fun x y => ?b + ?b(x := 1) + ?b(y := 1) + ?a(x) + ?a(x := y)
  let v := ?b(x := 4, y := 6)
  trace_state
  case b => exact 3 * x + y
  trace_state
  case a => exact x + 2
  trace_state
  trivial

/-!
Error: out of order
-/
example : True := by
  let f : Nat → Nat → Nat := fun x y => ?a
  let v := ?a(y := 1, x := 2)

/-!
Error: forward dependencies
-/
example : True := by
  let f : (n : Nat) → (m : Fin n) → Nat := fun n m => ?a
  let v := ?a(n := 1)

/-!
Using a metavariable later in a different context.
-/
example : (∀ x : Nat, x = x) ∧ (∀ x : Nat, x = x) := by
  refine ⟨fun x => ?a, ?_⟩
  · rfl
  · intro y
    exact ?a(x := y)

/-!
Using the metavariable later in a different context, in the same term.
-/
example : (∀ x : Nat, x = x) ∧ (∀ x : Nat, x = x) := by
  refine ⟨fun x => ?a(x), fun x => ?a(x)⟩
  · rfl
