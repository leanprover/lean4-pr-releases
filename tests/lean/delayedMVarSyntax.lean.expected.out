f : Nat → Nat := fun x => ?a(x := x)
g : Nat → Nat → Nat := fun x y => ?b(x := x, y := y) + ?b(y := y, x := 1) + ?b(x := x, y := 1) + ?a(x := x) + ?a(x := y)
v : Nat := ?b(x := 4, y := 6)
⊢ True

case b
f : Nat → Nat := fun x => ?a(x := x)
x y : Nat
⊢ Nat

case a
x : Nat
⊢ Nat
f : Nat → Nat := fun x => ?a(x := x)
g : Nat → Nat → Nat := fun x y => 3 * x + y + (3 * 1 + y) + (3 * x + 1) + ?a(x := x) + ?a(x := y)
v : Nat := 3 * 4 + 6
⊢ True

case a
x : Nat
⊢ Nat
f : Nat → Nat := fun x => x + 2
g : Nat → Nat → Nat := fun x y => 3 * x + y + (3 * 1 + y) + (3 * x + 1) + (x + 2) + (y + 2)
v : Nat := 3 * 4 + 6
⊢ True
delayedMVarSyntax.lean:24:11-24:29: error: local variables in delayed assignment syntax must appear in order according to the metavariable's local context:
case a
x y : Nat
⊢ Nat
delayedMVarSyntax.lean:22:18-24:29: error: unsolved goals
f : Nat → Nat → Nat := fun x y => ?a(x := x, y := y)
⊢ True
delayedMVarSyntax.lean:31:11-31:21: error: local variables in delayed assignment syntax must not have forward dependencies in the metavariable's local context:
case a
n : Nat
m : Fin n
⊢ Nat
delayedMVarSyntax.lean:29:18-31:21: error: unsolved goals
f : (n : Nat) → Fin n → Nat := fun n m => ?a(n := n, m := m)
⊢ True
